### ISSUE TITLE
A question on the implementation

### ISSUE THREAD

**Issue by wwang2** [gle-bellier/flow-matching#1]
First of all, thanks for this nice implementation!

I have a question on the implementation. According to the Lipman et al., the regression is performed on the conditional vector field $u(x|x_1), x \sim p_t(x|x_1)$, but your implementation is trying to match $v_t(x_0)$ with  $u(x_0 | x_1), x_0 \sim   p(x_0) $. Is this choice consistent with eqn (9) in Lipman et al?

**Comment by lebellig** [comment]
Hi!

Thank you for your feedback on the notebook!
About your question, indeed Eq.9 reminds us that:

$$
L_{CFM}(\theta) = \mathbb{E}_{t, q(x_1), p_t(x|x_1)} \lVert v_t(x) - u_t(x | x_1) \rVert^2
$$

and then you're right, we have to sample $x \sim p(x | x_1)$. However when dealing with Gaussian conditional probability paths (e.g. Eq.10: $p_t(x | x_1) = N(x | \mu_t(x_1), \sigma_t(x_1)^2I)$ )
we can consider flows as $\psi_t$ (Eq.11): 

$$
\psi_t(x) = \sigma_t(x_1)x + \mu_t(x_1)
$$

which satisfies (Eq.12):

$$
[\psi_t]_* p(x) = p_t(x | x_1)
$$

where $p(x)$ is our prior. Thus Eq.9 can be rewritten by remplacing $x \sim p(x | x_1)$ with $x_0 \sim p(x_0)$ and then applying $\psi$ on $x_0$:

$$
L_{CFM}(\theta) = \mathbb{E}_{t, q(x_1), p(x_0)} \lVert v_t(\psi_t(x_0)) - u_t(\psi_t(x_0) | x_1) \rVert^2
$$

The second term has a closed in the case of Optimal Transport conditional vector fields (Eq.13,21,22) and we obtain the final loss (Eq.23).
In the code it corresponds to:

```python
  def loss(self, v_t: nn.Module, x_0: torch.Tensor, x_1: torch.Tensor, t: torch.Tensor) -> torch.Tensor:
    """ Compute loss
    """ 
    v_psi = v_t(t, self.psi_t(x_0, x_1, t))
    d_psi = x_1 - (1 - self.sig_min) * x_0
    return torch.mean((v_psi - d_psi) ** 2
```
I hope it helps to understand why in the notebook I am not sampling directly from $p_t(x | x_1)$. I refactored the code in version 2 (available soon) to include sampling in the loss function and make it clearer.

**Comment by wwang2** [comment]
Thank you so much for the explanation! I think the full derivation for `d_psi` or $u_t(\phi_t (x_0 ) | x_1)$ should be the following: 

Based on eqn.12
 $$\phi_t(x) = [1 - (1-\sigma_{min}) ] x + t x_1$$  
so 
$$\phi_t(x_0) = [1 - (1-\sigma_{min}) ] x_0 + t x_1 $$

And from eqn. 21 

$$u_t(x|x_1) = \frac{x_1 - (1 - \sigma_{min} ) x }{ 1 - (1 - \sigma_{min}) t}$$ ,

Based on those, we have

$$
\begin{aligned}
u_t( \phi_t(x_0)  |x_1) &= \frac{x_1 - (1 - \sigma_{min} ) \phi_t(x_0) }{ 1 - (1 - \sigma_{min}) t}  \\
                                      &= \frac{x_1 - (1 - \sigma_{min} ) [[1 - (1-\sigma_{min}) ] x_0 + t x_1] }{ 1 - (1 - \sigma_{min}) t} \text{(with some term rearrangment)}\\
&=x_1 - (1 - \sigma_{min}) x_0
\end{aligned}
$$  


I think what confused me before was that $u_t( \phi_t(x_0)  |x_1)$ coincides with $u_0 (x_0 | x_1)$ based on equation 21, but it makes sense to me now because the vector field is constant for a straight conditional path. 

Another source confusion was to recognize $x$ is just $x_0$, I wonder if it makes more sense to use $x_0$ throughout the paper.

**Comment by wwang2** [comment]
Also, (maybe I am still not understanding completely), do you think it makes more sense to stick with the notation  $u_t(\psi_t(x) )| x_1)$ (eqn. 14) instead of $u_t(x | x_1)$ (eqn. 15)? simply because $\psi_0(x) = x$ is the initial condition $x_0$.

**Comment by lebellig** [comment]
I think you made a small typo and forgot a $t$ factor in the equation of the flow (Eq.22)

$$\psi_t(x) = (1 - (1 - \sigma_{min})t)x + tx_1 $$

It spread throughout the proof but once the modification is made yes it works.
Alternatively, Eq.13 (which holds for all the vector fields mentioned in the article) can be used and you only have to differentiate the flow:

$$u_t(\psi_t(x)|x_1) = \frac{d}{dt}\psi_t(x) $$

But I understand that working directly with the vector field allows underlining the fact this last is constant when considering Optimal Transport conditional VFs.

However, I'm not sure I understand your second comment. IMHO $x$ should be considered as a point of $\mathbb{R}^d$ that allows defining in all generality of the vector fields on the domain while the $x_0$ and $x_1$ notations allow evoking the trajectories and the flows. 
For the notebook, I preferred to keep the same notations in order to make it easier to read in parallel to the article.

**Comment by wwang2** [comment]
Thanks for catching the typo and thanks again for the explanation. Let me try to summarize your explanation

$$
\psi_t(x) : \text{conditional probability path with }  \psi_0(x) = x,   \psi_1(x) = x_1
$$

$$
u_t(x| x_1) : \text{vector field that defines the propagation of }  \psi_t(x) 
$$

$$
\frac{d \psi_t(x) }{ dt } = u_t(\psi_t(x) | x_1):  \text{ODE that generates the flow  }  \psi_t(x) 
$$

Practically, we are generally interested in $x_0 \sim p(x_0) = \mathcal{N} (x_0 | 0, I) $, but other $p(x_0)$ is also possible, as long as it is compatible with the choice of $\mu_0 (x_1)$ and $\sigma_0 (x_1) $.

Sorry if my question was too naive. Your explanation helped a lot!

**Comment by lebellig** [comment]
Yes, we can select other priors as long as we know how to sample them effectively and we can properly define the corresponding flow.
I'm glad if the notebook and the conversation are useful. I close this issue since the questions have been answered but feel free to keep in touch via email or Twitter!
